### PR TITLE
Add interfaces for services in AttributeBundle

### PR DESCRIPTION
**Pull request by jkrzefski**
### 1. Why is this change necessary?
Every service in the dependency injection container deserves to have an interface. It enables third party developers to decorate these services with ease.

### 2. What does this change do, exactly?
It adds new interfaces for the services in the AttributeBundle with all the public methods signatures those services already implement.

### 4. Please link to the relevant issues (if any).
None.

### 5. Which documentation changes (if any) need to be made because of this PR?
The page about the [attribute system](https://developers.shopware.com/developers-guide/attribute-system/) should receive an update to let people know about these new interfaces.

### 6. Checklist

- [ ] I have written tests and verified that they fail without my change
- [x] I have squashed any insignificant commits
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

### CODE REVIEW

File: engine/Shopware/Bundle/AttributeBundle/Service/CrudServiceInterface.php [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+ * our trademarks remain entirely with us.
+ */
+
+namespace Shopware\Bundle\AttributeBundle\Service;
+
+interface CrudServiceInterface
+{
+    /**
+     * @param string $table
+     * @param string $column
```
Comment (shyim) [inline engine/Shopware/Bundle/AttributeBundle/Service/CrudServiceInterface.php:34]: please move also the phpdocs

**Review by shyim — CHANGES_REQUESTED** [review]

**Comment by mitelg** [comment]
thanks for your PR @jkrzefski 
you also have to change the typehints in classes, where these classes are used. otherwise the interfaces will have no effect :wink:

**Comment by jkrzefski** [comment]
@mitelg Yes, I will do that.

File: engine/Shopware/Bundle/AttributeBundle/Service/CrudServiceInterface.php (hunk 1 above)
Comment (jkrzefski) [reply engine/Shopware/Bundle/AttributeBundle/Service/CrudServiceInterface.php:34]: Can you explain, what exactly do you mean?

**Comment by jkrzefski** [comment]
Btw: I left out the `@throws` statements of the phpdocs on purpose in the interfaces. The reason for me to not have those in the interfaces is, that I believe they are specific to the implementation. Do you approve this?

**Comment by mitelg** [comment]
@jkrzefski I would say it depends :wink: 
Some interfaces could describe certain throw tags, as the caller of the service will handle them for example. But I don't think that this is the case here

**Comment by jkrzefski** [comment]
@mitelg Ok, then I'll keep it this way. Should I change the phpdocs of the implementations to `{@inheritdoc}`?

File: engine/Shopware/Bundle/AttributeBundle/Service/CrudService.php [hunk 2]
```diff
@@ -27,36 +27,36 @@
 use Shopware\Components\Model\ModelManager;
 use Shopware\Models\Attribute\Configuration;
 
-class CrudService
+class CrudService implements CrudServiceInterface
 {
-    const EXT_JS_PREFIX = '__attribute_';
-    const NULL_STRING = 'NULL';
+    const EXT_JS_PREFIX = CrudServiceInterface::EXT_JS_PREFIX;
```
Comment (shyim) [inline engine/Shopware/Bundle/AttributeBundle/Service/CrudService.php]: Is this really necessary when the implemented interface contains that?
https://3v4l.org/rotPu

Comment (jkrzefski) [reply engine/Shopware/Bundle/AttributeBundle/Service/CrudService.php]: Thank you. I did not think about this. I'll remove it.

File: engine/Shopware/Bundle/AttributeBundle/Service/TypeMapping.php [hunk 3]
```diff
@@ -26,95 +26,95 @@
...
-    const TYPE_HTML = 'html';
-    const TYPE_INTEGER = 'integer';
-    const TYPE_FLOAT = 'float';
-    const TYPE_BOOLEAN = 'boolean';
-    const TYPE_DATE = 'date';
-    const TYPE_DATETIME = 'datetime';
-    const TYPE_COMBOBOX = 'combobox';
-    const TYPE_SINGLE_SELECTION = 'single_selection';
-    const TYPE_MULTI_SELECTION = 'multi_selection';
+    const TYPE_STRING = TypeMappingInterface::TYPE_STRING;
```
Comment (mitelg) [inline engine/Shopware/Bundle/AttributeBundle/Service/TypeMapping.php]: why the duplication?

Comment (shyim) [reply engine/Shopware/Bundle/AttributeBundle/Service/TypeMapping.php]: https://github.com/shopware/shopware/pull/2222#discussion_r331024163

**Comment by shyim** [comment]
Thanks for your Contribution! :blue_heart:

**Pull request closed without merging**